app/routes.py

```diff
@@ -1,2 +1,345 @@
+
+
+# define a new route to update (PUT) one task by its id (route parameter between <>, treated as a variable):
+@tasks_bp.route("/<task_id>", methods=["PUT"], strict_slashes=False)
+def update_single_task(task_id):
+
+    if not is_int(task_id):        
+        return {"message": f"ID {task_id} must be an integer"}, 400
+    
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+
+    form_data=request.get_json()
+    task.title=form_data["title"]
+    task.description=form_data["description"]
+    task.completed_at=form_data["completed_at"]
+    db.session.commit()
+
+    # we need an extra step to transfer completed_at to is_completed:
+    task_dict = task.to_dict()
+    
+    return jsonify({"task": task_dict}), 200
+
+
+# define a new route to DELETE one task by its id (route parameter between <>, treated as a variable):
+@tasks_bp.route("/<task_id>", methods=["DELETE"], strict_slashes=False)
+def delete_single_task(task_id):
+
+    if not is_int(task_id):        
+        return {"message": f"ID {task_id} must be an integer"}, 400
+
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+   
+    db.session.delete(task)
+    db.session.commit()
+
+    return jsonify({"details": f'Task {task.id} "{task.title}" successfully deleted'}), 200
+
+
+
+### WAVE 3 - PATCH ROUTES for TASKS
+
+# define a new route to update (PATCH) one task by its id (route parameter between <>, treated as a variable):
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"], strict_slashes=False)
+def mark_complete(task_id):
+
+    if not is_int(task_id):        
+        return {"message": f"ID {task_id} must be an integer"}, 400
+    
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+
+    task.completed_at=datetime.datetime.now()
+    post_message(task.title)
+    db.session.commit()
+
+    # we need an extra step to transfer completed_at to is_completed:
+    task_dict = task.to_dict()
+    
+    return jsonify({"task": task_dict}), 200
+
+# Wave 4 - Post message to Slack:
+def post_message(task_title):
+    requests.post("https://slack.com/api/chat.postMessage", \
+            headers={"Authorization": f"Bearer {os.environ.get('SLACK_TOKEN')}"}, \
+                data={"channel": f"{os.environ.get('SLACK_CHANNEL')}", "text":f"Someone just completed the task {task_title}"})
+
+
+
+# define a new route to update (PATCH) one task by its id (route parameter between <>, treated as a variable):
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"], strict_slashes=False)
+def mark_incomplete(task_id):
+
+    if not is_int(task_id):        
+        return {"message": f"ID {task_id} must be an integer"}, 400
+    
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+
+    if task.completed_at != None:
+        task.completed_at = None
+        task.is_complete = False
```

Comment:
I think you can take this line out as your `to_dict()` method already handles the value of `is_complete`.